drud/ddev
# drud-s3 provider implementation, fixes #998

## The Problem/Issue/Bug:

ddev-Live needs integration, uses s3 buckets

## How this PR Solves The Problem:

Implement drud-s3 provider plugin

Remaining work:
- [x] Some ugly specific mentions of provider type need to just be refactored. Can be done much better. 
- [x] AWS credentials need to be pulled into ~/.ddev, they're currently in the ~/.aws/credentials standard place. 
- [x] Bucket Name has to be moved into config.
- [x] Validate environment in ValidateField
- [x] Use https://github.com/AlecAivazis/survey for prompting at least here.
- [x] Validate environment in environmentPrompt()
- [x] I don't think the environment prompt is actually working right; no display of available environments.
- [x] Checking for an existing project/site should probably *probe by sitename-prefix* instead of getting all available keys
- [x] Consider changing import.yaml to drud-s3.yaml
- [x] Figure out how `ddev config drud-s3 --flag --flag` is supposed to work, this is ugly everywhere, in fact we added custom flags for pantheon.
- [x] Tests for the drud-s3 provider
- [x] Tests for the `ddev config drud-s3` command
- [x] Add to `ddev pull` tests to support the new provider.
- [x] Add docs for usage
- [ ] Make sure that AWS credentials used for tests do not allow access to writable directories.
- [x] Make tests for config UI work, or spin off issue to do that with survey

Review at this point should just be general "works/doesn't work" and "looks like the code is going the right/wrong direction", since there's quite a bit of work remaining


## Manual Testing Instructions:

* Clone one of the ddev-local-tests repositories, git info in the 1password ddev vault, item called "ddev-local/ddev-live development environments" 
* `ddev config drud-s3` - AWS credentials are in that same 1Password vault item.
* CHOOSE THE bucket ddev-local-tests - Unfortunately our AWS S3 keys allow us to see way too many buckets.
* Any environment that is available there should work for this, they all have backups in them.
* `ddev pull`
* Look at the local project site. 
* Experiment with command-line arguments `ddev config drud-s3 -h`
* Test regular config and `ddev config pantheon` usage. `ddev config` has been completely and totally refactored.

## Automated Testing Overview:

* Basic testing for the drud-s3 provider is in place
* Testing of the UI (prompting) is not there at all, since we couldn't figure out how to do it yet.

Needs much:
- [ ] Overall provider testing

## Related Issue Link(s):

OP #998 

## Release/Deployment notes:
